crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2497,71 +2497,98 @@ impl<'db> TypeInferenceBuilder<'db> {
             self.check_division_by_zero(binary, left_ty);
         }
 
+        self.infer_binary_expression_type(left_ty, right_ty, *op)
+            .unwrap_or_else(|| {
+                self.add_diagnostic(
+                    binary.into(),
+                    "unsupported-operator",
+                    format_args!(
+                        "Operator `{op}` is unsupported between objects of type `{}` and `{}`",
+                        left_ty.display(self.db),
+                        right_ty.display(self.db)
+                    ),
+                );
+                Type::Unknown
+            })
+    }
+
+    fn infer_binary_expression_type(
+        &mut self,
+        left_ty: Type<'db>,
+        right_ty: Type<'db>,
+        op: ast::Operator,
+    ) -> Option<Type<'db>> {
```

Comment:
@carljm -- currently I just returned an `Option`, as it's simpler and we don't have a branch dealing with `Union`s yet. When we implement `Union`s I figured we could refactor it to return a `Result` rather than an `Option`, but this seemed more minimal and easier to understand for now.

Reply:
I would probably have just gone for `Result` from the start, if we know that's where we'll end up, but I don't feel strongly, or that you need to change it. We can definitely just do it in the PR that adds union or tuple support.